// @frully/google-translate-open-api/dist/translate.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var token_1 = require("./token");
var rp = require("request-promise-native");
var util_1 = require("./util");
var language_1 = require("./language");
var constant_1 = require("./constant");
function handletranslate(data, extra) {
    return __awaiter(this, void 0, void 0, function () {
        var e, tk, query, options, browersUrl, res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (extra.from) {
                        if (!language_1.isSupport(extra.from)) {
                            e = new Error();
                            e.language = extra.from;
                        }
                    }
                    if (!language_1.isSupport(extra.to)) {
                        e = new Error();
                        e.language = extra.to;
                    }
                    if (e) {
                        e.code = 400;
                        e.message = 'The language \'' + e.language + '\' is not supported';
                        return [2 /*return*/, e];
                    }
                    return [4 /*yield*/, token_1.getTK(data.join(''), {
                            tld: extra.tld || 'com',
                            proxy: extra.proxy || false,
                            config: extra.config
                        })];
                case 1:
                    tk = _a.sent();
                    query = {
                        anno: 3,
                        client: "webapp",
                        format: 'html',
                        v: 1.0,
                        key: null,
                        logld: "vTE_20190506_00",
                        sl: extra.from || 'auto',
                        tl: extra.to || 'zh-CN',
                        hl: 'zh-CN',
                        sp: "nmt",
                        tc: 2,
                        sr: 1,
                        tk: tk,
                        mode: 1
                    };
                    options = __assign({ method: 'POST', headers: {}, form: util_1.arrayStringify(data), url: 'https://translate.googleapis.com/translate_a/t', qs: query, json: true }, (extra.config));
                    if (extra.proxy) {
                        options.proxy = "http://" + extra.proxy.host + ":" + extra.proxy.port;
                    }
                    options.headers = Object.assign({
                        "User-Agent": constant_1.userAgent,
                    }, options.headers, {
                        'content-type': 'application/x-www-form-urlencoded',
                        'Accept': 'application/json, text/plain, */*',
                        'X-Requested-With': 'XMLHttpRequest'
                    });
                    browersUrl = 'https://cors-anywhere.herokuapp.com/';
                    if (extra.browersUrl) {
                        browersUrl = extra.browersUrl;
                    }
                    if (extra.browers) {
                        options.baseURL = browersUrl + options.baseURL;
                    }
                    return [4 /*yield*/, rp(options)];
                case 2:
                    res = _a.sent();
                    return [2 /*return*/, {
                            data: res
                        }];
            }
        });
    });
}
exports.default = handletranslate;
